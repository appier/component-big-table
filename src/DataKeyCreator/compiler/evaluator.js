// evaluator accecpt an AST and variableMap
// and return evaulated value.

const evaluate = node => variable => {

  if(!node){
    throw `Expression is not valid`;
  }

  if(node.type === 'NumberLiteral'){
    return Number(node.value);
  }
  else if(node.type === 'Identifier'){
    //support Immutable
    let ret = null;
    if(typeof variable.get === 'function'){
      ret =  variable.get(node.value);
    }
    else{
      ret = variable[node.value];
    }

    if(ret){
      return ret;
    }
    throw `Unknown identifier: ${node.value}`;
  }
  else if(node.type === 'BinaryExpression'){
    if(node.operator === '+'){
      return evaluate(node.left)(variable) + evaluate(node.right)(variable);
    }
    else if(node.operator === '-'){
      return evaluate(node.left)(variable) - evaluate(node.right)(variable);
    }
    else if(node.operator === '*'){
      return evaluate(node.left)(variable) * evaluate(node.right)(variable);
    }
    else if(node.operator === '/'){
      return evaluate(node.left)(variable) / evaluate(node.right)(variable);
    }
  }
  else {
    throw `Unhandle AST Node: ${node.type}`;
  }
}

const evaluator = ast => variable => {
  return ast.body.map(d => {
    return evaluate(d)(variable);
  });
}

export default evaluator;

