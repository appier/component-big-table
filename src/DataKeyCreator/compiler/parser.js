const parser = (tokens) => {
  let position = 0;

  const additive = () => {
    let left = multiplicative();
    if(position < tokens.length){
      let token = tokens[position];
      // console.log('additive', position, token)
      while ( token && (token.value === '+' || token.value === '-' ) ) {
        position++;
        left = {
          type: 'BinaryExpression',
          left: left,
          operator: token.value,
          right: multiplicative()
        };
        token = tokens[position];
      }
    }
    return left;
  };

  const multiplicative = () => {
    let left = primay();
    if(position < tokens.length){
      let token = tokens[position];
      // console.log('multiplicative', position, token)
      while ( token && (token.value === '*' || token.value === '/') ) {
        position++;
        left = {
          type: 'BinaryExpression',
          operator: token.value,
          left: left,
          right: primay(),
        };
        token = tokens[position];
      }
    }
    return left;
  }

  const primay = () => {
    if(position < tokens.length){
      let token = tokens[position];
      let ret = null;
      // console.log('primay', position, token)
      if ( token.value === '(' ) {
        position++;
        ret = additive();
        token = tokens[position];
        // consume ')'
        if(token && token.value === ')') {
          position++;
        }
        else {
          throw 'Unmatched paren: it seems you lack a ) in expression';
        }
      }
      else if ( token.type === 'number' ) {
        position++;
        ret = {
          type: 'NumberLiteral',
          value: token.value,
        };
      }
      else if ( token.type === 'identifier' ) {
        position++;
        ret = {
          type: 'Identifier',
          value: token.value,
        };
      }
      else {
        throw 'Unexpected token: ' + token.value;
      }
      return ret;
    }
  }

  let body = [];
  while (position < tokens.length) {
    body.push(additive());
  }

  let ast = {
    type: 'Program',
    body: body,
  };

  return ast;
}

export default parser;