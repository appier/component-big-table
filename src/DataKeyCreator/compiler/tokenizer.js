
const WHITESPACE_TEST = /\s/;
const NUMBERS_TEST = /[0-9]/;
const OPERATOR_TEST = /[\-+*/]/;
const IDENTIFIER_HEAD_TEST = /[a-z]/i;
const IDENTIFIER_TEST = /[a-z0-9_\-]/i;

const tokenizer = (input) => {

  let position = 0;
  let tokens = [];

  while(position < input.length){
    let char = input[position];

    if (char === '(') {
      tokens.push({
        type: 'paren',
        value: '(',
      });
      position++;
      continue;
    }

    if (char === ')') {
      tokens.push({
        type: 'paren',
        value: ')',
      });
      position++;
      continue;
    }

    //ignore white space..
    if (WHITESPACE_TEST.test(char)) {
      position++;
      continue;
    }

    //abstract identifier
    if (IDENTIFIER_HEAD_TEST.test(char)) {
      let value = '';
      while (
        (IDENTIFIER_TEST.test(char) || NUMBERS_TEST.test(char))
        && position < input.length
      ) {
        value += char;
        char = input[++position];
      }
      tokens.push({
        type: 'identifier',
        value: value
      });
      continue;
    }

    //abstract number
    if(NUMBERS_TEST.test(char)) {
      let value = '';
      while (NUMBERS_TEST.test(char) && position < input.length) {
        value += char;
        char = input[++position];
      }
      tokens.push({
        type: 'number',
        value: value
      });
      continue;
    }

    //abstract operator
    if(OPERATOR_TEST.test(char)) {
      let value = '';
      while (OPERATOR_TEST.test(char) && position < input.length) {
        value += char;
        char = input[++position];
        if(value.length >= 2){
          throw new TypeError('illegal operator: '+ value);
        }
      }
      tokens.push({
        type: 'operator',
        value: value
      });
      continue;
    }


    throw new TypeError('Unable to parse character: ' + char);
  }

  return tokens;
};

export default tokenizer;