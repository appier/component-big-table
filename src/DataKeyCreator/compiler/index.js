import tokenizer from './tokenizer';
import parser from './parser';
import evaluator from './evaluator';

export const compiler = expr => variable => {

  try {
    const tokens = tokenizer(expr);
    const ast = parser(tokens);
    const values = evaluator(ast)(variable);
    if(values.length === 1 && values[0]){
      return {
        success: true,
        value: values[0],
      }
    }
    return {
      success: false,
      err: 'Expression is not valid',
    }
  } catch(err) {
    return {
      success: false,
      err: String(err),
    }
  }

}




export default compiler;