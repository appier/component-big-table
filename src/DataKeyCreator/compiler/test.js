import tokenizer from './tokenizer';
import parser from './parser';
import evaluator from './evaluator';


console.log(tokenizer('a b +'));

let ast = null;
let value = null;
// ast = parser([
//   { type: 'identifier', value: 'cost' }
// ]);
// console.log(JSON.stringify(ast, null, 2));

// ast = parser([
//   { type: 'identifier', value: 'action' },
//   { type: 'operator', value: '*' },
//   { type: 'identifier', value: 'cost' }
// ]);
// console.log(JSON.stringify(ast, null, 2));


// ast = parser([
//   { type: 'number', value: '3' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '2' }
// ]);
// console.log(JSON.stringify(ast, null, 2));


// ast = parser([
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '*' },
//   { type: 'number', value: '2' },
//   { type: 'operator', value: '*' },
//   { type: 'number', value: '3' }
// ]);
// console.log(JSON.stringify(ast, null, 2));



// ast = parser([
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '*' },
//   { type: 'number', value: '2' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '3' },
//   { type: 'operator', value: '*' },
//   { type: 'number', value: '4' },
// ]);
// console.log(JSON.stringify(ast, null, 2));



// ast = parser([
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '2' },
//   { type: 'operator', value: '*' },
//   { type: 'number', value: '3' },
//   { type: 'operator', value: '-' },
//   { type: 'number', value: '4' },
//   { type: 'operator', value: '/' },
//   { type: 'number', value: '5' }
// ]);
// console.log(JSON.stringify(ast, null, 2));


// ast = parser([
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '*' },
//   { type: 'paren', value: '(' },
//   { type: 'number', value: '2' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '3' },
//   { type: 'paren', value: ')' },
// ]);
// console.log(JSON.stringify(ast, null, 2));

// ast = parser([
//   { type: 'paren', value: '(' },
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '2' },
//   { type: 'paren', value: ')' },
//   { type: 'operator', value: '*' },
//   { type: 'paren', value: '(' },
//   { type: 'number', value: '2' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '3' },
//   { type: 'paren', value: ')' },
// ]);
// console.log(JSON.stringify(ast, null, 2));

// ast = parser([
//   { type: 'number', value: '1' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '2' },
// ]);
// console.log(JSON.stringify(ast, null, 2));


ast = parser([
  { type: 'identifier', value: 'a' },
  { type: 'identifier', value: 'b' },
  { type: 'operator', value: '+' }
]);
console.log(JSON.stringify(ast, null, 2));


// ast = parser([
//   { type: 'number', value: '3' },
//   { type: 'operator', value: '+' },
//   { type: 'number', value: '2' }
// ]);

// ast = parser([
//     { type: 'paren', value: '(' },
//     { type: 'identifier', value: 'impression' },
//     { type: 'operator', value: '+' },
//     { type: 'identifier', value: 'click' },
//     { type: 'paren', value: ')' },
//     { type: 'operator', value: '/' },
//     { type: 'paren', value: '(' },
//     { type: 'identifier', value: 'cost' },
//     { type: 'operator', value: '+' },
//     { type: 'identifier', value: 'action' },
//     { type: 'paren', value: ')' },
// ]);
// value = evaluator(ast)({
//   cost: 100,
//   impression: 6000,
//   click: 300,
//   action: 2,
// })


// console.log(JSON.stringify(ast, null, 2));
// console.log('---------------')
// console.log('value: ', value)
// console.log('---------------')



