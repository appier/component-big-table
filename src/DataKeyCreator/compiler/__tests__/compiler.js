import compiler from '../index';
import { fromJS } from 'immutable';

const data = {
  cost: 100,
  impression: 6000,
  click: 300,
  action: 2,
};

const immutableData = fromJS(data);

describe('Compiler', () => {

  it('Empty', () => {
    expect(compiler('')(data)).toMatchSnapshot();
  })

  it('Number', () => {
    expect(compiler('1')(data)).toMatchSnapshot();
  })

  it('Identifier', () => {
    expect(compiler('cost')(data)).toMatchSnapshot();
  })

  it('BinaryExpression: Number', () => {
    expect(compiler('1+2')(data)).toMatchSnapshot();
  })

  it('BinaryExpression: Number, Identifier', () => {
    expect(compiler('1+cost')(data)).toMatchSnapshot();
  })

  it('BinaryExpression: Identifier, Identifier', () => {
    expect(compiler('action/cost')(data)).toMatchSnapshot();
  })

  it('Paren: Number', () => {
    expect(compiler('(1+2)')(data)).toMatchSnapshot();
  })

  it('Paren: Number 2', () => {
    expect(compiler('(1+2)*3')(data)).toMatchSnapshot();
  })

  it('Paren: Number, Identifier', () => {
    expect(compiler('(1+action) / cost')(data)).toMatchSnapshot();
  })

  it('Support Immutable', () => {
    expect(compiler('(1+action) / cost')(immutableData)).toMatchSnapshot();
  })

  it('Support Immutable 2', () => {
    expect(compiler(' ( 1+action) / (cost - 20)')(immutableData)).toMatchSnapshot();
  })

  it('Error Handlering: tokenizer error', () => {
    expect(compiler('1 && 2')(data)).toMatchSnapshot();
  })

  it('Error Handlering: tokenizer error 2', () => {
    expect(compiler('1 + 2 )')(data)).toMatchSnapshot();
  })

  it('Error Handlering: tokenizer error 3', () => {
    expect(compiler('action cost +')(data)).toMatchSnapshot();
  })

  it('Error Handlering: parser error', () => {
    expect(compiler('((1+2)')(data)).toMatchSnapshot();
  })

  it('Error Handlering: evaluator error', () => {
    expect(compiler('1 + aa')(data)).toMatchSnapshot();
  })


});