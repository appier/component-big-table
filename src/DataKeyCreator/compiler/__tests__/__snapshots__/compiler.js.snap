exports[`Compiler BinaryExpression: Identifier, Identifier 1`] = `
Object {
  "success": true,
  "value": 0.02,
}
`;

exports[`Compiler BinaryExpression: Number 1`] = `
Object {
  "success": true,
  "value": 3,
}
`;

exports[`Compiler BinaryExpression: Number, Identifier 1`] = `
Object {
  "success": true,
  "value": 101,
}
`;

exports[`Compiler Empty 1`] = `
Object {
  "err": "Expression is not valid",
  "success": false,
}
`;

exports[`Compiler Error Handlering: evaluator error 1`] = `
Object {
  "err": "Unknown identifier: aa",
  "success": false,
}
`;

exports[`Compiler Error Handlering: parser error 1`] = `
Object {
  "err": "Unmatched paren: it seems you lack a ) in expression",
  "success": false,
}
`;

exports[`Compiler Error Handlering: tokenizer error 1`] = `
Object {
  "err": "TypeError: Unable to parse character: &",
  "success": false,
}
`;

exports[`Compiler Error Handlering: tokenizer error 2 1`] = `
Object {
  "err": "Unexpected token: )",
  "success": false,
}
`;

exports[`Compiler Error Handlering: tokenizer error 3 1`] = `
Object {
  "err": "Expression is not valid",
  "success": false,
}
`;

exports[`Compiler Identifier 1`] = `
Object {
  "success": true,
  "value": 100,
}
`;

exports[`Compiler Number 1`] = `
Object {
  "success": true,
  "value": 1,
}
`;

exports[`Compiler Paren: Number 1`] = `
Object {
  "success": true,
  "value": 3,
}
`;

exports[`Compiler Paren: Number 2 1`] = `
Object {
  "success": true,
  "value": 9,
}
`;

exports[`Compiler Paren: Number, Identifier 1`] = `
Object {
  "success": true,
  "value": 0.03,
}
`;

exports[`Compiler Support Immutable 1`] = `
Object {
  "success": true,
  "value": 0.03,
}
`;

exports[`Compiler Support Immutable 2 1`] = `
Object {
  "success": true,
  "value": 0.0375,
}
`;
